### PR TITLE
collect columns for merge

### PR DESCRIPTION
**Pull request by PSeitz**

### CODE REVIEW

File: columnar/src/column_index/mod.rs [hunk 1]
```diff
@@ -44,9 +44,10 @@ impl<'a> ColumnIndex<'a> {
                 }
             }
             ColumnIndex::Multivalued(multivalued_index) => {
-                let start = multivalued_index.get_val(row_id);
-                let end = multivalued_index.get_val(row_id + 1);
-                start..end
+                unimplemented!()
+                // let start = multivalued_index.get_val(row_id);
```
Comment (fulmicoton) [inline columnar/src/column_index/mod.rs]: Why unimplemented!? Did you spot a bug?

File: columnar/src/columnar/merge.rs [hunk 2]
```diff
@@ -31,3 +35,135 @@ pub fn merge_columnar(
...
+        }
+    }
+}
+
+/// Receives a list of columns of numerical types (u64, i64, f64)
+///
+/// Returns a list of `DynamicColumn` which are all of the same numerical type
+fn cast_to_common_numerical_column(columns: &[DynamicColumn]) -> Vec<DynamicColumn> {
+    assert!(columns.iter().all(|column| column.is_numerical()));
+    let coerce_to_i64: Vec<_> = columns
```
Comment (fulmicoton) [inline columnar/src/columnar/merge.rs:86]: we can probably do something a little bit nicer here, and be more coupled with the writer logic?

```rust
fn acceptable_numerical_type(numerical_types: &[NumericalType]) -> NumericalType {
     let mut compatible_numerical_types = CompatibleNumericalTypes::default();
     for &typ in numerical_types {
         compatible_numerical_types.accept(typ);
     }
     numerical_types.to_numeric_type()
}


struct DynamicColummn {
    fn column_type(&self) -> ColumnType { ... }
    fn coerce(&self, column_type: ColumnType) -> Option<DynamicColumn> { ... }
}
```

File: columnar/src/dynamic_column.rs [hunk 3]
```diff
@@ -20,6 +22,114 @@ pub enum DynamicColumn {
...
+    }
+    pub fn is_f64(&self) -> bool {
+        matches!(self, DynamicColumn::F64(_))
+    }
+    pub fn is_u64(&self) -> bool {
+        matches!(self, DynamicColumn::U64(_))
+    }
+    pub fn is_i64(&self) -> bool {
+        matches!(self, DynamicColumn::I64(_))
+    }
```
Comment (fulmicoton) [inline columnar/src/dynamic_column.rs]: I'd replace those by a `.column_type()`...
That one already have a conversion to numerical_type()

File: columnar/src/dynamic_column.rs [hunk 4]
```diff
@@ -20,6 +22,114 @@ pub enum DynamicColumn {
...
+                    values: Arc::new(monotonic_map_column(column.values, MapI64ToU64)),
+                }))
+            }
+            DynamicColumn::U64(_) => Some(self),
+            _ => None,
+        }
+    }
+}
+
+pub struct MapI64ToF64;
```
Comment (fulmicoton) [inline columnar/src/dynamic_column.rs]: Do you think we could replace those three thing by the following?
I don;'t know any trait boudnary for as.

If you want to stick to `as`, can you make those private to this module?
Technically these mapping are not monotonic.



```
pub struct Coerce<FromTyp, ToTyp>;

impl<FromTyp, ToTyp> StrictlyMonotonicFn<FromTyp, ToTyp> for MapI64ToF64 
where FromTyp: TryFrom<ToTyp>, ToTyp: TryFrom<FromTyp>
{
}
```

**Review by fulmicoton — CHANGES_REQUESTED** [review]
see comments inline

File: columnar/src/columnar/merge.rs (hunk 2 above)
Comment (PSeitz) [reply columnar/src/columnar/merge.rs:86]: the type is not enough, we need to know if the a i64 column could be a u64 column (!contains negative values)

File: columnar/src/dynamic_column.rs (hunk 4 above)
Comment (PSeitz) [reply columnar/src/dynamic_column.rs]: maybe with num traits, but the use case is too small

**Review by fulmicoton — APPROVED** [review]

File: columnar/src/dynamic_column.rs (hunk 4 above)
Comment (fulmicoton) [reply columnar/src/dynamic_column.rs]: ok

File: columnar/src/columnar/merge.rs (hunk 2 above)
Comment (fulmicoton) [reply columnar/src/columnar/merge.rs:86]: ah good point

**Pull request merged**